### DISCUSSION TITLE
rtp stream through input.ffmpeg

### DISCUSSION THREAD

**Discussion by Russsgithub** [savonet/liquidsoap#2318] [User Support]
Liq v. 2.1

Has anyone had any luck streaming multiple rtp streams into liq via input.ffmpeg.

I've been trying this 

```
input.ffmpeg(url:port?localaddr=xxx.xxx.xxx.xxx&fifo_size=49152&overrun_nonfatal=1&reuse=1)
```

localaddr should allow it to listen on the local ip for an rtp stream.

One instance of this seems to connect ok but more than one hangs the server.

I'm guessing there is a reason that an rtp implementation with ffmpeg has not been added since input.udp was depreciated in v2.0.

Thanks

**Comment by toots** [comment]
Hi!

What happens if you do the same with two different calls to the ffmpeg command line?

**Reply by Russsgithub** [reply]
I think i got the wrong end of the stick in the comment below..

I'll have to rearange stuff to test tthe he liq script from two ffmpeg sources.

I have noticed that running one through this script

```
s = input.ffmpeg(start=false, int_args=[("reorder_queue_size", 100)], format="rtp", "rtp://127.0.0.1:50000?localaddr=127.0.0.1&reuse=1&fifo_size=49152&overrun_nonfatal=1")

s.start()

print(s.status())

thread.run(every=1., fun() -> print(s.status()))

s = switch(track_sensitive= false, transition_length=1., [({true}, s),({true}, blank())])

output.dummy(id="2", fallible=true, s)
```

That the script sits waiting on running, printing "pending" . When I make the rtp stream available it connects and stream, printing connected. When i drop the rtp stream there is a timeout disconection and the input happily sits waiting for the stream to come back.

If I add another rtp input

```
s = input.ffmpeg(start=false, int_args=[("reorder_queue_size", 100)], format="rtp", "rtp://127.0.0.1:50000?localaddr=127.0.0.1&reuse=1&fifo_size=49152&overrun_nonfatal=1")
x = input.ffmpeg(start=false, int_args=[("reorder_queue_size", 100)], format="rtp", "rtp://127.0.0.2:60000?localaddr=127.0.0.1&reuse=1&fifo_size=49152&overrun_nonfatal=1")
s.start()
x.start()
print(s.status())

thread.run(every=1., fun() -> print(s.status()))

s = switch(track_sensitive= false, transition_length=1., [({true}, s),({true}, blank())])

output.dummy(id="2", fallible=true, s)
```

I get no printing of pending as in the first script. If i send the rtp stream all works.

I think this may be where my script is hanging.

**Reply by Russsgithub** [reply]
If i print the status of x every second as well nothing is logged until I start the input stream, when the status of both is printed until the input stream is dropped, then the status of neither is printed again.

**Reply by Russsgithub** [reply]
```
s = input.ffmpeg(start=true, int_args=[("reorder_queue_size", 100)], format="rtp", "rtp://127.0.0.1:50000?localaddr=127.0.0.1&reuse=1&fifo_size=49152&overrun_nonfatal=1")
x = input.ffmpeg(start=true, int_args=[("reorder_queue_size", 100)], format="rtp", "rtp://127.0.0.2:60000?localaddr=127.0.0.1&reuse=1&fifo_size=49152&overrun_nonfatal=1")

thread.run(every=1., fun() -> print(s.status()))
thread.run(every=1., fun() -> print(x.status()))

output.dummy(id="2", fallible=true, s)
output.dummy(id="1", fallible=true, x)
```

**Reply by Russsgithub** [reply]
```
connected rtp://127.0.0.1:50000?localaddr=127.0.0.1&reuse=1&fifo_size=49152&overrun_nonfatal=1
connected rtp://127.0.0.1:50000?localaddr=127.0.0.1&reuse=1&fifo_size=49152&overrun_nonfatal=1
polling
polling
```
I get a double hit on each in the logs every second with this script when one has something to connect too.

**Reply by Russsgithub** [reply]
If I use sdp files
```
s = input.ffmpeg(int_args=[("analyzeduration", 2147483647),("probesize", 2147483647),("connect", 1)],string_args=[("protocol_whitelist","file,crypto,udp,rtp")],"./test.sdp")
x = input.ffmpeg(int_args=[("analyzeduration", 2147483647),("probesize", 2147483647),("connect", 1)],string_args=[("protocol_whitelist","file,crypto,udp,rtp")],"./test2.sdp")
```

with different input ip i get
```
[udp @ 0x7fdc64009280] bind failed: Address already in use
[udp @ 0x7fdc6803b0c0] bind failed: Address already in use
```
Until I send one stream, when all works as expected.
i.e. alternating connect to , and polling in log.

netcat tells me that the relavent ports are bound to 0.0.0.0 rather than the ips used in the sdp files, 127.0.0.1 and an external ip I can't put here.

**Comment by Russsgithub** [comment]
Hi @toots and @smimram 

I'm still not getting this to behave,. Having set up a similar situation with cli ffmpeg processes running on the same computer I find that it works differently to liq. My situation is this

I have now managed to get a script working that doesn't lock up my server, but still has the underlying issue.
To try and simplify things.

I have a dynamic source server which can accept http and rtp ( though input.ffmpeg() ) and outputs the source to icecast.

If I start an http source all works well.

If I start an unconnected (the source url is not producing a stream) mksafe(rtp) source it creates the icecast output, then if I start an http source all works well.

The problem -

If I create 2 unconnected rtp sources and then an http source the 2 rtp sources create their icecast outputs, but the http source doesn't and therefore doesn't stream.

I've tried this in ffmpeg cli and with 2 unconnected rtp inputs waiting to stream to dev/null, where I can successfully connect an http stream to an icecast output.

I haven't tried this with the rtp streams trying to output to  icecast, which i guess is next on my list.

**Reply by Russsgithub** [reply]
I've just tried this with rtp to icecast in ffmpeg, and can confirm there are no issues. 

i.e. 2 x unconnected rtp input to icecast. http input to icecast works.

Also connecting the rtp stream works.

Something doesn't seem to be working , but it may be me.

**Comment by Russsgithub** [comment]
I made this work by increasing the scheduler generic queues.

I guess that input.ffmpeg with a rtp/udp source that has nothing to stream blocks the queue until it can stream.
